signac/synced_collections/backends/collection_mongodb.py

```diff
@@ -0,0 +1,187 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Implements a MongoDB SyncedCollection backend."""
+from .. import SyncedAttrDict, SyncedCollection, SyncedList
+
+try:
+    import bson
+
+    MONGO = True
+except ImportError:
+    MONGO = False
+
+
+class MongoDBCollection(SyncedCollection):
+    """A :class:`SyncedCollection` that synchronizes with a MongoDB document.
+
+    MongoDB stores documents in collections, where each document is a single
+    record encoded in a JSON-like format (BSON). Each :class:`SyncedCollection` can be
+    represented as a MongoDB document, so this backend stores the :class:`SyncedCollection`
+    as a single document within the collection provided by the user. The document
+    is identified by a unique key provided by the user.
+
+    **Thread safety**
+
+    The MongoDBCollection is not thread-safe.
+
+    Parameters
+    ----------
+    collection : :py:class:`pymongo.collection.Collection`
+        The MongoDB client in which to store data.
+    uid : dict
+        The unique key-value mapping added to the data and stored in the document
+        so that it is uniquely identifiable in the MongoDB collection.
+
+    Warnings
+    --------
+    The user is responsible for providing a unique id such that there are no
+    possible collisions between different :class:`SyncedCollection` instances
+    stored in the same MongoDB collection. Failure to do so may result data
+    corruption if multiple documents are found to be apparently associated with
+    a given :class:`SyncedCollection`.
+
+    """
+
+    _backend = __name__  # type: ignore
+
+    def __init__(self, collection=None, uid=None, parent=None, **kwargs):
+        if not MONGO:
+            raise RuntimeError(
+                "The PyMongo package must be installed to use the MongoDBCollection."
+            )
+
+        self._collection = collection
+        self._uid = uid
+        super().__init__(parent=parent, **kwargs)
```

Comment:
I don't understand how this MongoDB data model is supposed to work with nested synced collections. I don't think I'm following how the `collection` / `uid` are used in conjunction with the `parent` to write the correct data to the correct Mongo document.